eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java

```diff
@@ -905,13 +905,41 @@ public T detect(Predicate<? super T> predicate)
     }
 
     @Override
+    public <P> T detectWith(Predicate2<? super T, ? super P> predicate, P parameter)
+    {
+        return InternalArrayIterate.detectWith(this.items, this.size, predicate, parameter);
+    }
+
+    @Override
+    public Optional<T> detectOptional(Predicate<? super T> predicate)
+    {
+        return InternalArrayIterate.detectOptional(this.items, this.size, predicate);
+    }
+
+    @Override
+    public <P> Optional<T> detectWithOptional(Predicate2<? super T, ? super P> predicate, P parameter)
+    {
+        return InternalArrayIterate.detectWithOptional(this.items, this.size, predicate, parameter);
+    }
+
+    @Override
     public T detectIfNone(Predicate<? super T> predicate, Function0<? extends T> defaultValueBlock)
     {
         T result = this.detect(predicate);
         return result == null ? defaultValueBlock.value() : result;
     }
 
     @Override
+    public <P> T detectWithIfNone(
+            Predicate2<? super T, ? super P> predicate,
+            P parameter,
+            Function0<? extends T> defaultValueBlock)
+    {
+        T result = this.detectWith(predicate, parameter);
+        return result == null ? defaultValueBlock.value() : result;
+    }
+
```

Comment:
There is one interesting edge case where this doesn't work.

```java
Assert.assertEquals(Integer.valueOf(3), this.newWith(Comparators.reverseNaturalOrder(), 2, 3, 4, 5).detectIfNone(Integer.valueOf(3)::equals, null));
```

When it detects an element that meets the predicate condition, it should return it even if the function passed is null. However it throws NullPointerException with the method reference approach above (it seems method reference is "evaluated" even though it is not executed in the code path). 

We can change it as below to address this particular case still using detectOptional().

```java
this.detectOptional(predicate).orElseGet(() -> (function == null) ? null : function.value());
```

However now this code behaves differently from the original code when it detects null AND null function is passed in. Original code throws NPE, while the code above returns null.

I think this behaviour was not documented or tested anywhere, but I suppose original code behaves more naturally in this case (throws NPE). 

So I'd revert this to be original null check code. Let me know what you think.